Improve keyboard interaction of the DropdownMenu

afercia opened the pull request:
This PR tries to improve the accessibility and keyboard interaction of the DropdownMenu following the model described in the ARIA Authoring Practices.

The DropdownMenu is, in ARIA terms, a composite widget made of a [Menu button](https://www.w3.org/TR/wai-aria-practices/#menubutton) and a [Menu](https://www.w3.org/TR/wai-aria-practices/#menu). To get how the expected keyboard interaction should work I'd recommend to jump directly to the W3C examples for the standalone Menu:
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions.html
and the same example in the context of a Toolbar:
https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html

Navigation through the menu items should work with arrow keys only, while tabbing should close the menu (same for Escape). Optionally, items navigation should "loop" continuously. 

Worth reminding ARIA widgets replicate common, established patterns already used in our operating systems and keyboard users are used to them. For example, the macOS Finder menu works this way: it uses arrow keys to navigate the drop-down items:

<img width="326" alt="screen shot 2017-07-21 at 19 17 36" src="https://user-images.githubusercontent.com/1682452/28476310-e994f6e8-6e4f-11e7-8ef9-bd40d2ea9313.png">

The mixed use of the Tab key and arrow keys is a great way to reduce the amount of Tab key presses necessary to navigate through content, as it allows to open a menu just if that's really needed.

As mentioned in the issue, it would be great to consider to extend this behavior to other components too.

Fixes #1880

afercia committed:
ARIA menu keyboard interaction first pass.

afercia committed:
Focus the first item on open and some clean-up.

afercia committed:
Reduce usage of findDOMNode and some clean-up.

afercia commented:
Fixes #1880

aduth commented on components/dropdown-menu/index.js:
```diff
@@ -60,47 +79,58 @@ class DropdownMenu extends wp.element.Component {
-		if ( this.menuRef ) {
-			const menu = findDOMNode( this.menuRef );
+		if ( this.menuNode ) {
 			if ( index < 0 ) {
```
With changes to `focusPrevious`, is it ever the case that this condition would pass? Also generally concerned about `previousElementSibling` potentially not being a focusable element. Might be worth dropping this.

afercia replied:
OK

aduth commented on components/dropdown-menu/index.js:
```diff
@@ -47,10 +50,26 @@ class DropdownMenu extends wp.element.Component {
+
+		this.maybeCloseMenuOnBlur = setTimeout( () => {
+			this.closeMenu();
+		}, 100 );
```
How did you settle on `100` ?

afercia replied:
Completely arbitrary (I know...). There's a brief interval between blur on an element and focus on the following element. '0' didn't work. On the W3C examples, they use '300', see:
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/js/Menubutton.js
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/js/MenuItemAction.js
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/js/PopupMenuAction.js

Open to suggestions...

aduth commented on components/dropdown-menu/index.js:
```diff
@@ -60,47 +79,58 @@ class DropdownMenu extends wp.element.Component {
-					this.closeMenu();
 					const node = findDOMNode( this );
 					const toggle = node.querySelector( '.components-dropdown-menu__toggle' );
 					toggle.focus();
```
I'm finding that occasionally pressing escape when navigating buttons will clear focus altogether (`document.activeElement === document.body`). Not sure if that's specific to your changes here though.

macOS Chrome 59.0.3071.115

afercia replied:
Same chrome here and can't reproduce. Updated to 60: same.
Instead, I think `handleKeyUp()` doesn't work as intended: when I keep Escape pressed for a while, the menu closes immediately so when I release Escape the condition to stop propagation isn't met and the toolbar disappears.

aduth commented on components/dropdown-menu/index.js:
```diff
@@ -192,7 +219,10 @@ class DropdownMenu extends wp.element.Component {
+						ref={ this.bindMenuReferenceNode }
+						onFocus={ this.handleFocus }
+						onBlur={ this.handleBlur }
+						tabIndex="-1"
```
Is `tabindex` necessary on `div`?

afercia committed:
Use keyup to make Escape close the menu and simplify all the things.

afercia commented:
Using keyup to close the menu when pressing Escape allows to greatly simplify. No timers, no focus/blur. To recap:
- menu items can be navigated with arrows only
- tabbing will close the menu and natively focus the next/previous focusable elements (note: there's a pre-existing and unrelated bug when using Firefox, will open a separate issue)
- pressing Escape closes the menu and focuses the toggle
- one more Escape press makes the whole toolbar disappear, as expected

aduth commented on components/dropdown-menu/index.js:
```diff
@@ -60,62 +61,62 @@ class DropdownMenu extends wp.element.Component {
+		if ( event.keyCode === ESCAPE && this.state.open ) {
+			event.preventDefault();
+			event.stopPropagation();
+			const node = findDOMNode( this );
```
We could probably create a separate `ref` for the root rendered element to avoid the `findDOMNode` here.

aduth commented on components/dropdown-menu/index.js:
```diff
@@ -60,62 +61,62 @@ class DropdownMenu extends wp.element.Component {
+			event.preventDefault();
+			event.stopPropagation();
+			const node = findDOMNode( this );
+			const toggle = node.querySelector( '.components-dropdown-menu__toggle' );
```
Actually, for that matter, we could bind the `ref` to the toggle to avoid the `querySelector` as well.

aduth commented:
Would be good to have some unit tests for this component, but that can be addressed separately I think.

aduth committed:
Target toggle focus by ref

aduth commented:
In e91f3a2, I revised `IconButton` to be a stateful component class, so that we can apply a `ref` to target it directly shifting focus to the toggle when the menu closes. Let me know if this looks good to you, otherwise I think this is ready to merge.

aduth committed:
Fix function binding for node references

aduth committed:
Find toggle node to focus

When used on a component class, the value of a ref is the component instance, not the DOM node.

https://facebook.github.io/react/docs/refs-and-the-dom.html#adding-a-ref-to-a-class-component

afercia committed:
Use Component instead of wp globals.
